### PR TITLE
First changes to structure and AlexNet integration

### PR DESCRIPTION
**Pull request by fuzhanrahmanian**
**Describe your improvements**
Added AlexNet model with Keras.
Modified model.py
Reshape output file structure.
Added main and argpars for ease of use
Updated .gitignore to ignore .h5 files

### CODE REVIEW

File: featurevis/featurevis.py [hunk 1]
```diff
@@ -131,7 +132,7 @@ def visualize_filter(image, model, layer, filter_index, iterations,
     print('>> 100 %')
     # Decode the resulting input image
     image = imgs.deprocess_image(image[0].numpy())
-    return loss, image
+    return loss, image, layer, filter_index
```
Comment (Sparkier) [inline featurevis/featurevis.py:135]: Why would we return layer and filter_index? We already provide this to the function as a parameter, so the other end should know about this, no?

File: featurevis/image_reader.py [hunk 2]
```diff
@@ -15,7 +16,7 @@ def save_npy_as_png(path):
     """
     filepath = Path(path)
     filename = filepath.stem
-    out_path = Path("{0}.png".format(filename))
+    out_path = Path("outputs/{0}.png".format(filename))
```
Comment (Sparkier) [inline featurevis/image_reader.py:19]: I think this should not be changed this way. With the current implementation of taking a path and putting the file there, the user can precisely decide where he wants the output to go. If we prefix with `output/`, then it always needs to be put in a folder called output, which I think is not necessary.

File: main.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# model.summary()
+image = images.initialize_image(224, 224)
+loss, image, layer_name, channel_num = featurevis.visualize_filter(image, model, "mixed6", 1, 500, 2, 0, 0, 0)
+
+name = "feature_vis_{}_{}_{}".format(arch, layer_name, channel_num)
+print(loss)
+
+images.save_image(image, name= name)
+image_reader.save_npy_as_png("{}.npy".format(name))
+
```
Comment (Sparkier) [inline main.py:23]: I think we should not have a `main.py` like this. If people are supposed to use this as a library (as is probably intended), i.e., to write their own code and import this as a package (`import luna`), then such a file is not needed. What we could think of is two options:

1. Currently, we have a how-to demo in the README.md. It could stay this way and we could change that readme if we think this code is a better showcase.
2. We could do a `luna_demo.py`, which does what you have here. The downside of this is that when the project is used as a library, this does not really make sense, as there is no method to call.

These things considered, I would move this code to the README.md, if we think this is a better showcase. What do you think?

File: pretrained_models/alexnet.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
+"""
+A Keras Implementation of the GoogleNet (InceptionV1) from
+https://github.com/heuritech/convnets-keras/blob/master/convnetskeras/convnets.py
```
Comment (Sparkier) [inline pretrained_models/alexnet.py:3]: This comment is incorrect.

File: pretrained_models/alexnet.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    if weights_path:
+        model.load_weights(weights_path)
+
+    return model
+
+
+if __name__ == "__main__":
+    # Test pretrained model
+    model = AlexNet('alexnet_weights.h5')
+    model.summary()
```
Comment (Sparkier) [inline pretrained_models/alexnet.py:187]: I think we should remove that, one could always load the model by importing it from pretrained models and then do the summary. We should not have `main` stuff in this library IMO.

File: pretrained_models/models.py [hunk 6]
```diff
@@ -20,10 +21,16 @@ def get_model(model_name):
...
         tf.compat.v1.keras.backend.set_image_data_format('channels_first')
         return googlenet.create_googlenet()
     if model_name == "inceptionV1slim":
         inputs = tf.random.uniform((1, 224, 224, 3), dtype=tf.dtypes.float32)
         return slim_nets.inception.inception_v1(inputs)
-    return keras.applications.VGG16(weights="imagenet", include_top=False)
+    if model_name == "VGG16":
+        return keras.applications.vgg16(weights= "imagenet", include_top=False)
+    if model_name == "alexnet":
+        return alexnet.AlexNet()
```
Comment (Sparkier) [inline pretrained_models/models.py:33]: This does not use a `weights_path`, is that intended? For other models, we load the pretrained weights by default.

File: pretrained_models/models.py [hunk 7]
```diff
@@ -20,10 +21,16 @@ def get_model(model_name):
...
+    else:
+        print("Model is not specified")
```
Comment (Sparkier) [inline pretrained_models/models.py:35]: The problem about this is, that we do have one path now that does not return anything. I put a default return here that returns VGG16 if nothing else matches. Maybe we should go back to that and print a warning? Otherwise the code will fail if the wrong model is specified. In the long run, we might want to have individual functions for the models to get rid of this `if/else`, which is not very pretty.

File: featurevis/featurevis.py [hunk 8]
```diff
@@ -1,10 +1,11 @@
 """
 The main file for the feature vis process
 """
+
 import tensorflow as tf
 from tensorflow import keras
 
-from luna.featurevis import images as imgs
+from featurevis import images as imgs
```
Comment (Sparkier) [inline featurevis/featurevis.py:8]: Out of curiosity, why did you change that? I think the previous version was correct, no? At least if you use this as a library.

**Review by Sparkier — CHANGES_REQUESTED** [review]
Thank you for the PR, its a good starter and especially the amount of changes is nice as it is not too large to review :)

I added some comments which we should discuss before merging. You can obviously already start more work on another branch which can be a follow-up PR.

Additionally, there are a bunch of binary files which I think should not be committed, those are:
`feature_vis_inceptionV3_mixed6_1.npy`
`outputs/feature_vis_inceptionV3_mixed6_1.png`
`pretrained_models/data/meta_clsloc.mat`
`test.npy`
`test.png`

I think in general, you want to have the `main.py` outside of this project, and all the output as well.

File: main.py (hunk 3 above)
Comment (fuzhanrahmanian) [reply main.py:23]: I would agree more to put it inside README.md. Maybe we can talk more about it in our meeting.

File: pretrained_models/alexnet.py (hunk 4 above)
Comment (fuzhanrahmanian) [reply pretrained_models/alexnet.py:3]: You are right. I wrote the name of architecture wrong. 

```suggestion
A Keras Implementation for Alexnet from
https://github.com/heuritech/convnets-keras/blob/master/convnetskeras/convnets.py
```

File: pretrained_models/alexnet.py (hunk 5 above)
Comment (fuzhanrahmanian) [reply pretrained_models/alexnet.py:187]: Ok, I will make the change!

File: pretrained_models/models.py (hunk 6 above)
Comment (fuzhanrahmanian) [reply pretrained_models/models.py:33]: I wrote the initialization of the weight_path inside the main of alexnet.py. But if we want to get rid of it, I will give the pretrained weight as default in the model.py file.

File: pretrained_models/models.py (hunk 7 above)
Comment (fuzhanrahmanian) [reply pretrained_models/models.py:35]: We can talk about it. I just don't like the VGG as fallback but we can decide about this.

File: featurevis/featurevis.py (hunk 8 above)
Comment (fuzhanrahmanian) [reply featurevis/featurevis.py:8]: I did this because I wrote my main.py file inside the library and this luna was redundant at that point.

File: featurevis/image_reader.py (hunk 2 above)
Comment (fuzhanrahmanian) [reply featurevis/image_reader.py:19]: Probably it would be better for future users. I added this for having an order for myself since I always save the images in my repo folder. I will make the output path freely choosable.

File: featurevis/featurevis.py (hunk 1 above)
Comment (fuzhanrahmanian) [reply featurevis/featurevis.py]: You are right. I will implement the variables in the main function and use them for the file name generation there.

File: featurevis/featurevis.py (hunk 8 above)
Comment (Sparkier) [reply featurevis/featurevis.py:8]: In that case, we should probably change it back.

File: featurevis/image_reader.py [hunk 9]
```diff
@@ -2,20 +2,21 @@
...
 them as pngs in the output folder
 """
+
 from pathlib import Path
 import numpy as np
 from tensorflow import keras
 
 
-def save_npy_as_png(path):
+def save_npy_as_png(input_path, output_path = "lune/outputs"):
```
Comment (Sparkier) [inline featurevis/image_reader.py:11]: In general, I think this change is good. But I don't think we should have an outputs folder inside the project. If this is to be used as a library, this does not really make sense. So I would say this should probably not have a default value.

File: pretrained_models/models.py (hunk 7 above)
Comment (Sparkier) [reply pretrained_models/models.py:35]: Sure, lets talk about this tomorrow.

File: main.py (hunk 3 above)
Comment (Sparkier) [reply main.py]: Sure, lets talk tomorrow.

File: featurevis/featurevis.py [hunk 10]
```diff
@@ -132,6 +134,9 @@ def visualize_filter(image, model, layer, filter_index, iterations,
     print('>> 100 %')
     # Decode the resulting input image
     image = imgs.deprocess_image(image[0].numpy())
+    print(image.shape)
+    print(image)
+    # conver the proceed image to a valid rgb color
```
Comment (Sparkier) [inline featurevis/featurevis.py:139]: conver -> convert

File: featurevis/featurevis.py [hunk 11]
```diff
@@ -132,6 +134,9 @@ def visualize_filter(image, model, layer, filter_index, iterations,
     print('>> 100 %')
     # Decode the resulting input image
     image = imgs.deprocess_image(image[0].numpy())
+    print(image.shape)
+    print(image)
```
Comment (Sparkier) [inline featurevis/featurevis.py:138]: What is that for?

File: featurevis/featurevis.py [hunk 12]
```diff
@@ -175,10 +180,11 @@ def gradient_ascent_step(img, model, filter_index, learning_rate, channels_first
     with tf.GradientTape() as tape:
         tape.watch(img)
         loss = compute_loss(img, model, filter_index, channels_first)
+        #loss_prop = tf.reduce_sum(img, axis=-1)
```
Comment (Sparkier) [inline featurevis/featurevis.py:183]: Uncommented lines should be removed.

File: featurevis/featurevis.py [hunk 13]
```diff
@@ -175,10 +180,11 @@ def gradient_ascent_step(img, model, filter_index, learning_rate, channels_first
...
     # Compute gradients.
     grads = tape.gradient(loss, img)
     # Normalize gradients.
-    grads = tf.math.l2_normalize(grads)
+    #grads = tf.math.l2_normalize(grads)
```
Comment (Sparkier) [inline featurevis/featurevis.py:187]: Uncommented lines should be removed.

File: featurevis/image_reader.py [hunk 14]
```diff
@@ -8,7 +8,7 @@
 from tensorflow import keras
 
 
-def save_npy_as_png(input_path, output_path = "lune/outputs"):
+def save_npy_as_png(input_path, output_path = "luna/outputs"):
```
Comment (Sparkier) [inline featurevis/image_reader.py:11]: Still think per default, this should not be in `luna`.

**Review by Sparkier — APPROVED** [review]
